lib/model/indexhandler.go

```diff
@@ -222,15 +222,35 @@ func (s *indexHandler) pause() {
 // sendIndexTo sends file infos with a sequence number higher than prevSequence and
 // returns the highest sent sequence number.
 func (s *indexHandler) sendIndexTo(ctx context.Context, fset *db.FileSet) error {
+	// Keep track of the previous sequence we sent. This is separate from
+	// s.prevSequence because the latter will skip over holes in the
+	// sequence numberings, while sentPrevSequence should always be
+	// precisely the highest previousely sent sequence.
+	sentPrevSequence := s.prevSequence
+
 	initial := s.prevSequence == 0
 	batch := db.NewFileInfoBatch(nil)
 	batch.SetFlushFunc(func(fs []protocol.FileInfo) error {
+		var lastSequence int64
+		if len(fs) > 0 {
+			lastSequence = fs[len(fs)-1].Sequence
+		}
 		l.Debugf("%v: Sending %d files (<%d bytes)", s, len(fs), batch.Size())
 		if initial {
 			initial = false
-			return s.conn.Index(ctx, &protocol.Index{Folder: s.folder, Files: fs})
+			return s.conn.Index(ctx, &protocol.Index{
+				Folder:       s.folder,
+				Files:        fs,
+				LastSequence: lastSequence,
+			})
 		}
-		return s.conn.IndexUpdate(ctx, &protocol.IndexUpdate{Folder: s.folder, Files: fs})
+		defer func() { sentPrevSequence = lastSequence }()
```

Comment:
This is getting a little confusing to me.  You're setting a callback via `SetFlushFunc`.  It will execute this little helper whenever it exits here branch.  That last action is to modify a variable that's local to the outer function which installs the callback (`sendIndexTo`).  When will this flush function be called?  Is the call to `sendIndexTo` already completed by then?  If so, which local variable will get overwritten?

Reply:
It's called so sentPrevSequence is updated after s.conn.IndexUpdate completes but before the function returns. Essentially it could have been this

```
err := s.conn.IndexUpdate(ctx, &protocol.IndexUpdate{
	Folder:       s.folder,
	Files:        fs,
	PrevSequence: sentPrevSequence,
	LastSequence: lastSequence,
})
sentPrevSequence = lastSequence
return err
```

but that's ugly 🤷

Reply:
Well that part (what the defer does) is clear, but you didn't answer my actual questions. :wink:

Reply:
Maybe I don't understand the question then. If you're asking me to explain the entire mechanism of the existing batch/flush mechanism I think that's better if you just take a look at it. As for the "which local variable" question, if you mean sentPrevSequence that we're talking about, there is only one.

Reply:
The question was about timing.  Setting a callback function that accesses this local variable `sentPrevSequence` only works if the callback *invocation* happens before the variable's owner `sendIndexTo` has returned.  If that were true, then why is a callback needed instead of calling the callback function directly?

Reply:
Yeah, I can't actually parse your question. If by callback you mean the flush function, which I think is the only thing that resembles that description, it's called by batch.Flush() and (maybe) batch.Append().

> Setting a callback function that accesses this local variable sentPrevSequence only works if the callback invocation happens before the variable's owner sendIndexTo has returned. 

I'm not certain what precisely you mean by this, and I don't think it applies in this case anyhow as sendIndexTo() doesn't return until all flushes are complete, but even if it did it's perfectly fine to access a variable in a closure after the function that created the closure returns. This is very common pattern, even.

Reply:
> as sendIndexTo() doesn't return until all flushes are complete

That's the answer. Thank you.

"Flush function", "closure" or callback - yes I meant that same thing.  If `SetFlushFunc` stored a reference to that closure somewhere (now or in some future modification), there could be a reference to a function which writes to a variable that has long gone out of scope.  But I'm not sure storing a closure for later (like a function pointer or a "callable" in Python) is even possible in Go.  Anyway, we don't do it, so we're fine.

Reply:
It's not only possible, it happens every time you `go func(){ ... }()`, and any variables bound to the closure will share it's lifetime even if they're otherwise out of scope. 👍